PDK_Abstraction/FinFET14nm_Mock_PDK/Align_primitives.py

```diff
@@ -32,14 +23,39 @@ def main( args):
         x_cells = 2*args.Xcells
         # TODO: Fix difficulties associated with CC patterns matching this condition
         pattern = 2 if x_cells%4 != 0 else args.pattern ### CC is not possible; default is interdigitated
+    return x_cells, pattern
+
+def get_parameters(args):
+    parameters = {}
+    if args.model is None:
+        parameters['model'] = 'NMOS' if 'NMOS' in args.primitive else 'PMOS'
+    else:
+        parameters['model'] = args.model
+    parameters['width'] = '{width}' if args.width is None else args.width
+    parameters['length'] = '{length}' if args.length is None else args.length
+    parameters['nfin'] = args.nfin
+    return parameters
+
+def main( args):
+
+    logging.basicConfig(level=logging.getLevelName(args.logLevel))
+
+    fin = args.height
+    gateDummy = 3 ### Total Dummy gates per unit cell: 2*gateDummy
+    finDummy = 4  ### Total Dummy fins per unit cell: 2*finDummy
+    gate = 2
+    y_cells = args.Ycells
+
+    x_cells, pattern = get_xcells_pattern(args)
+    parameters = get_parameters(args)
 
     uc = primitive.PrimitiveGenerator( fin, finDummy, gate, gateDummy)
 
     def gen( pattern, routing):
         if 'NMOS' in args.primitive:
-            uc.addNMOSArray( x_cells, y_cells, pattern, routing)
+            uc.addNMOSArray( x_cells, y_cells, pattern, routing, **parameters)
```

Comment:
Haven't used this ** before. If parameters has a keyword argument that is already used, what happens? Here perhaps "pattern".

Reply:
The **parameters only catches keyword arguments that the function definition doesn't use already. So uc.addNMOSArray( x_cells, y_cells, pattern=pattern, routing=routing, **parameters) does not put pattern or routing in parameters.